File: src/services/bitsaccoApi.ts
/**
 * Bitsacco API Service Module for WhatsApp Assistant
 * TypeScript version with enhanced type safety
 * Handles integration with https://bitsacco.com/ backend API
 * 
 * @author Bitsacco Engineering Team
 * @version 2.0.0 - TypeScript Migration
 * 
 * Note: This is currently a placeholder service with mock implementations.
 * Replace mock logic with real HTTP requests to Bitsacco backend when available.
 */

import axios, { AxiosResponse } from 'axios';
import config from '../config/index.js';
import logger from '../utils/logger.js';

interface OtpResponse {
  success: boolean;
  message: string;
  otpId?: string;
  expiresAt?: string;
}

interface VerifyOtpResponse {
  success: boolean;
  userId?: string;
  message: string;
  token?: string;
  profile?: UserProfile;
}

interface UserProfile {
  userId: string;
  phone: string;
  name?: string;
  email?: string;
  kycStatus: string;
  accountStatus: string;
  createdAt: string;
}

interface BalanceResponse {
  success: boolean;
  balance: number;
  currency: string;
  btcBalance?: number;
  usdBalance?: number;
  message?: string;
}

interface LoadMoneyResponse {
  success: boolean;
  message: string;
  transactionId?: string;
  paymentInstructions?: string;
  qrCode?: string;
  amount: number;
  currency: string;
}

interface WithdrawResponse {
  success: boolean;
  message: string;
  transactionId?: string;
  estimatedArrival?: string;
  fees?: number;
}

interface Transaction {
  transactionId: string;
  type: 'deposit' | 'withdrawal' | 'transfer' | 'bitcoin_purchase' | 'bitcoin_sale';
  amount: number;
  currency: string;
  btcAmount?: number;
  date: string;
  status: 'pending' | 'completed' | 'failed' | 'cancelled';
  description?: string;
  fees?: number;
  destination?: string;
  source?: string;
}

interface TransactionHistoryResponse {
  success: boolean;
  transactions: Transaction[];
  totalCount: number;
  page: number;
  limit: number;
  message?: string;
}

interface PurchaseBitcoinResponse {
  success: boolean;
  message: string;
  transactionId?: string;
  btcAmount?: number;
  kesAmount: number;
  exchangeRate?: number;
  fees?: number;
}

interface WithdrawBitcoinResponse {
  success: boolean;
  message: string;
  transactionId?: string;
  btcAmount: number;
  kesAmount?: number;
  exchangeRate?: number;
  fees?: number;
}

interface ApiError {
  success: false;
  message: string;
  code?: string;
  details?: any;
}

class TypedBitsaccoApiService {
  private baseUrl: string;
  private apiKey: string;
  private timeout: number = 30000;
  private retryAttempts: number = 3;
  private retryDelay: number = 1000;

  constructor() {
    this.baseUrl = config.bitsacco?.apiUrl || 'https://api.bitsacco.com/v1';
    this.apiKey = config.bitsacco?.apiKey || '';
    
    if (!this.apiKey) {
      logger.warn('Bitsacco API key not configured - using mock responses (TypeScript)');
    }
    
    logger.info('Bitsacco API Service initialized (TypeScript)', {
      baseUrl: this.baseUrl,
      hasApiKey: Boolean(this.apiKey)
    });
  }

  /**
   * Send OTP to user's phone number
   */
  async sendOtp(phone: string): Promise<OtpResponse> {
    try {
      logger.info('Sending OTP (TypeScript)', { phone: this.maskPhone(phone) });

      // Mock implementation - replace with real API call
      if (!this.apiKey) {
        return this.mockSendOtp(phone);
      }

      const response: AxiosResponse<OtpResponse> = await this.makeApiCall('POST', '/auth/send-otp', {
        phone: phone,
        purpose: 'whatsapp_verification'
      });

      logger.info('OTP sent successfully (TypeScript)', {
        phone: this.maskPhone(phone),
        otpId: response.data.otpId
      });

      return response.data;

    } catch (error: any) {
      logger.error('Failed to send OTP (TypeScript)', {
        phone: this.maskPhone(phone),
        error: error.message
      });

      return {
        success: false,
        message: 'Failed to send OTP. Please try again.'
      };
    }
  }

  /**
   * Verify OTP for a phone number
   */
  async verifyOtp(phone: string, otp: string): Promise<VerifyOtpResponse> {
    try {
      logger.info('Verifying OTP (TypeScript)', { 
        phone: this.maskPhone(phone),
        otpLength: otp.length
      });

      // Mock implementation - replace with real API call
      if (!this.apiKey) {
        return this.mockVerifyOtp(phone, otp);
      }

      const response: AxiosResponse<VerifyOtpResponse> = await this.makeApiCall('POST', '/auth/verify-otp', {
        phone: phone,
        otp: otp
      });

      logger.info('OTP verified successfully (TypeScript)', {
        phone: this.maskPhone(phone),
        userId: response.data.userId
      });

      return response.data;

    } catch (error: any) {
      logger.error('Failed to verify OTP (TypeScript)', {
        phone: this.maskPhone(phone),
        error: error.message
      });

      return {
        success: false,
        message: 'Invalid OTP. Please try again.'
      };
    }
  }

  /**
   * Get Bitsacco account balance for a user
   */
  async getBalance(userId: string): Promise<BalanceResponse> {
    try {
      logger.info('Fetching balance (TypeScript)', { userId });

      // Mock implementation - replace with real API call
      if (!this.apiKey) {
        return this.mockGetBalance(userId);
      }

      const response: AxiosResponse<BalanceResponse> = await this.makeApiCall('GET', `/users/${userId}/balance`);

      logger.info('Balance fetched successfully (TypeScript)', {
        userId,
        balance: response.data.balance,
        currency: response.data.currency
      });

      return response.data;

    } catch (error: any) {
      logger.error('Failed to fetch balance (TypeScript)', {
        userId,
        error: error.message
      });

      return {
        success: false,
        balance: 0,
        currency: 'KES',
        message: 'Failed to fetch balance. Please try again.'
      };
    }
  }

  /**
   * Initiate a load money (deposit) transaction
   */
  async loadMoney(userId: string, amount: number, method: string = 'mpesa'): Promise<LoadMoneyResponse> {
    try {
      logger.info('Initiating deposit (TypeScript)', { userId, amount, method });

      // Mock implementation - replace with real API call
      if (!this.apiKey) {
        return this.mockLoadMoney(userId, amount, method);
      }

      const response: AxiosResponse<LoadMoneyResponse> = await this.makeApiCall('POST', `/users/${userId}/deposit`, {
        amount: amount,
        method: method,
        currency: 'KES'
      });

      logger.info('Deposit initiated successfully (TypeScript)', {
        userId,
        amount,
        method,
        transactionId: response.data.transactionId
      });

      return response.data;

    } catch (error: any) {
      logger.error('Failed to initiate deposit (TypeScript)', {
        userId,
        amount,
        method,
        error: error.message
      });

      return {
        success: false,
        message: 'Failed to initiate deposit. Please try again.',
        amount,
        currency: 'KES'
      };
    }
  }

  /**
   * Initiate a withdrawal transaction
   */
  async withdraw(userId: string, amount: number, destination: string, otp: string): Promise<WithdrawResponse> {
    try {
      logger.info('Processing withdrawal (TypeScript)', {
        userId,
        amount,
        destination: this.maskDestination(destination)
      });

      // Mock implementation - replace with real API call
      if (!this.apiKey) {
        return this.mockWithdraw(userId, amount, destination, otp);
      }

      const response: AxiosResponse<WithdrawResponse> = await this.makeApiCall('POST', `/users/${userId}/withdraw`, {
        amount: amount,
        destination: destination,
        otp: otp,
        currency: 'KES'
      });

      logger.info('Withdrawal processed successfully (TypeScript)', {
        userId,
        amount,
        transactionId: response.data.transactionId
      });

      return response.data;

    } catch (error: any) {
      logger.error('Failed to process withdrawal (TypeScript)', {
        userId,
        amount,
        error: error.message
      });

      return {
        success: false,
        message: 'Failed to process withdrawal. Please check your OTP and try again.'
      };
    }
  }

  /**
   * Purchase Bitcoin with KES
   */
  async purchaseBitcoin(userId: string, kesAmount: number): Promise<PurchaseBitcoinResponse> {
    try {
      logger.info('Purchasing Bitcoin (TypeScript)', { userId, kesAmount });

      // Mock implementation - replace with real API call
      if (!this.apiKey) {
        return this.mockPurchaseBitcoin(userId, kesAmount);
      }

      const response: AxiosResponse<PurchaseBitcoinResponse> = await this.makeApiCall('POST', `/users/${userId}/bitcoin/buy`, {
        amount: kesAmount,
        currency: 'KES'
      });

      logger.info('Bitcoin purchase successful (TypeScript)', {
        userId,
        kesAmount,
        btcAmount: response.data.btcAmount,
        transactionId: response.data.transactionId
      });

      return response.data;

    } catch (error: any) {
      logger.error('Failed to purchase Bitcoin (TypeScript)', {
        userId,
        kesAmount,
        error: error.message
      });

      return {
        success: false,
        message: 'Failed to purchase Bitcoin. Please try again.',
        kesAmount
      };
    }
  }

  /**
   * Withdraw Bitcoin for KES
   */
  async withdrawBitcoin(userId: string, btcAmount: number, otp: string): Promise<WithdrawBitcoinResponse> {
    try {
      logger.info('Withdrawing Bitcoin (TypeScript)', { userId, btcAmount });

      // Mock implementation - replace with real API call
      if (!this.apiKey) {
        return this.mockWithdrawBitcoin(userId, btcAmount, otp);
      }

      const response: AxiosResponse<WithdrawBitcoinResponse> = await this.makeApiCall('POST', `/users/${userId}/bitcoin/withdraw`, {
        btcAmount: btcAmount,
        otp: otp
      });

      logger.info('Bitcoin sale successful (TypeScript)', {
        userId,
        btcAmount,
        kesAmount: response.data.kesAmount,
        transactionId: response.data.transactionId
      });

      return response.data;

    } catch (error: any) {
      logger.error('Failed to withdraw Bitcoin (TypeScript)', {
        userId,
        btcAmount,
        error: error.message
      });

      return {
        success: false,
        message: 'Failed to withdraw Bitcoin. Please check your OTP and try again.',
        btcAmount
      };
    }
  }

  /**
   * Get recent transaction history for a user
   */
  async getTransactionHistory(userId: string, page: number = 1, limit: number = 10): Promise<TransactionHistoryResponse> {
    try {
      logger.info('Fetching transaction history (TypeScript)', { userId, page, limit });

      // Mock implementation - replace with real API call
      if (!this.apiKey) {
        return this.mockGetTransactionHistory(userId, page, limit);
      }

      const response: AxiosResponse<TransactionHistoryResponse> = await this.makeApiCall('GET', `/users/${userId}/transactions`, {
        page,
        limit
      });

      logger.info('Transaction history fetched successfully (TypeScript)', {
        userId,
        transactionCount: response.data.transactions.length,
        totalCount: response.data.totalCount
      });

      return response.data;

    } catch (error: any) {
      logger.error('Failed to fetch transaction history (TypeScript)', {
        userId,
        error: error.message
      });

      return {
        success: false,
        transactions: [],
        totalCount: 0,
        page,
        limit,
        message: 'Failed to fetch transaction history. Please try again.'
      };
    }
  }

  /**
   * Get user profile information
   */
  async getUserProfile(userId: string): Promise<UserProfile | null> {
    try {
      logger.info('Fetching user profile (TypeScript)', { userId });

      // Mock implementation - replace with real API call
      if (!this.apiKey) {
        return this.mockGetUserProfile(userId);
      }

      const response: AxiosResponse<{ success: boolean; profile: UserProfile }> = await this.makeApiCall('GET', `/users/${userId}/profile`);

      if (response.data.success) {
        logger.info('User profile fetched successfully (TypeScript)', {
          userId,
          kycStatus: response.data.profile.kycStatus
        });
        return response.data.profile;
      }

      return null;

    } catch (error: any) {
      logger.error('Failed to fetch user profile (TypeScript)', {
        userId,
        error: error.message
      });
      return null;
    }
  }

  /**
   * Make HTTP API call with retry logic
   */
  private async makeApiCall(method: string, endpoint: string, data?: any, params?: any): Promise<AxiosResponse<any>> {
    const url = `${this.baseUrl}${endpoint}`;
    const headers = {
      'Authorization': `Bearer ${this.apiKey}`,
      'Content-Type': 'application/json',
      'User-Agent': 'Bitsacco-WhatsApp-Bot/1.0'
    };

    const config = {
      method,
      url,
      headers,
      timeout: this.timeout,
      data: method !== 'GET' ? data : undefined,
      params: method === 'GET' ? data : params
    };

    for (let attempt = 1; attempt <= this.retryAttempts; attempt++) {
      try {
        const response = await axios(config);
        return response;
      } catch (error: any) {
        if (attempt === this.retryAttempts || !this.shouldRetry(error)) {
          throw error;
        }
        
        await new Promise(resolve => setTimeout(resolve, this.retryDelay * attempt));
      }
    }

    throw new Error('Max retry attempts reached');
  }

  /**
   * Determine if error should trigger a retry
   */
  private shouldRetry(error: any): boolean {
    return (
      error.code === 'ECONNRESET' ||
      error.code === 'ETIMEDOUT' ||
      (error.response && error.response.status >= 500)
    );
  }

  /**
   * Mask phone number for logging
   */
  private maskPhone(phone: string): string {
    if (phone.length <= 4) return phone;
    return `${phone.slice(0, 4)}****${phone.slice(-2)}`;
  }

  /**
   * Mask destination for logging
   */
  private maskDestination(destination: string): string {
    if (destination.length <= 4) return destination;
    return `${destination.slice(0, 4)}****${destination.slice(-2)}`;
  }

  // Mock implementations (remove when real API is available)
  private mockSendOtp(phone: string): OtpResponse {
    return {
      success: true,
      message: `OTP sent to ${this.maskPhone(phone)}`,
      otpId: `mock-otp-${Date.now()}`,
      expiresAt: new Date(Date.now() + 5 * 60 * 1000).toISOString()
    };
  }

  private mockVerifyOtp(phone: string, otp: string): VerifyOtpResponse {
    if (otp === '123456' || otp === '000000') {
      return {
        success: true,
        userId: `user-${phone.slice(-4)}`,
        message: 'OTP verified successfully',
        token: `mock-token-${Date.now()}`,
        profile: {
          userId: `user-${phone.slice(-4)}`,
          phone,
          kycStatus: 'pending',
          accountStatus: 'active',
          createdAt: new Date().toISOString()
        }
      };
    }
    return { success: false, message: 'Invalid OTP' };
  }

  private mockGetBalance(userId: string): BalanceResponse {
    return {
      success: true,
      balance: 5000 + Math.floor(Math.random() * 10000),
      currency: 'KES',
      btcBalance: 0.001 + Math.random() * 0.01,
      usdBalance: 50 + Math.random() * 100
    };
  }

  private mockLoadMoney(userId: string, amount: number, method: string): LoadMoneyResponse {
    return {
      success: true,
      message: `Deposit of KES ${amount} via ${method} initiated`,
      transactionId: `dep-${Date.now()}`,
      paymentInstructions: method === 'mpesa' 
        ? 'Send to PayBill 123456, Account: your-phone-number'
        : 'Bank transfer instructions will be sent via SMS',
      amount,
      currency: 'KES'
    };
  }

  private mockWithdraw(userId: string, amount: number, destination: string, otp: string): WithdrawResponse {
    if (otp === '654321' || otp === '000000') {
      return {
        success: true,
        message: `Withdrawal of KES ${amount} to ${this.maskDestination(destination)} successful`,
        transactionId: `wit-${Date.now()}`,
        estimatedArrival: '2-5 minutes',
        fees: amount * 0.01
      };
    }
    return { success: false, message: 'Invalid OTP for withdrawal' };
  }

  private mockPurchaseBitcoin(userId: string, kesAmount: number): PurchaseBitcoinResponse {
    const btcAmount = kesAmount / 8775000; // Mock BTC/KES rate
    const fees = kesAmount * 0.015; // 1.5% fee
    
    return {
      success: true,
      message: `Successfully purchased ₿${btcAmount.toFixed(8)} for KES ${kesAmount}`,
      transactionId: `btc-save-${Date.now()}`,
      btcAmount,
      kesAmount,
      exchangeRate: 8775000,
      fees
    };
  }

  private mockWithdrawBitcoin(userId: string, btcAmount: number, otp: string): WithdrawBitcoinResponse {
    if (otp === '654321' || otp === '000000') {
      const kesAmount = btcAmount * 8750000; // Mock BTC/KES rate (slightly lower for sell)
      const fees = kesAmount * 0.015; // 1.5% fee
      
      return {
        success: true,
        message: `Successfully sold ₿${btcAmount} for KES ${kesAmount.toFixed(2)}`,
        transactionId: `btc-withdraw-${Date.now()}`,
        btcAmount,
        kesAmount: kesAmount - fees,
        exchangeRate: 8750000,
        fees
      };
    }
    return { success: false, message: 'Invalid OTP for Bitcoin sale', btcAmount };
  }

  private mockGetTransactionHistory(userId: string, page: number, limit: number): TransactionHistoryResponse {
    const mockTransactions: Transaction[] = [
      {
        transactionId: 'tx-001',
        type: 'deposit',
        amount: 2000,
        currency: 'KES',
        date: new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString(),
        status: 'completed',
        description: 'M-Pesa deposit',
        source: '254700123456'
      },
      {
        transactionId: 'tx-002',
        type: 'bitcoin_purchase',
        amount: 1500,
        currency: 'KES',
        btcAmount: 0.0001709,
        date: new Date(Date.now() - 12 * 60 * 60 * 1000).toISOString(),
        status: 'completed',
        description: 'Bitcoin purchase',
        fees: 22.5
      },
      {
        transactionId: 'tx-003',
        type: 'deposit',
        amount: 4000,
        currency: 'KES',
        date: new Date(Date.now() - 2 * 60 * 60 * 1000).toISOString(),
        status: 'pending',
        description: 'Bank transfer deposit',
        source: 'KCB Bank'
      }
    ];

    return {
      success: true,
      transactions: mockTransactions.slice((page - 1) * limit, page * limit),
      totalCount: mockTransactions.length,
      page,
      limit
    };
  }

  private mockGetUserProfile(userId: string): UserProfile {
    return {
      userId,
      phone: '+254700123456',
      name: 'John Doe',
      kycStatus: 'verified',
      accountStatus: 'active',
      createdAt: new Date(Date.now() - 30 * 24 * 60 * 60 * 1000).toISOString()
    };
  }
}

// Export singleton instance
export const typedBitsaccoApiService = new TypedBitsaccoApiService();
export default typedBitsaccoApiService;

File: app/services/bitsacco_api.py
"""
Bitsacco API Client - Production-ready integration
Handles all API communication with Bitsacco.com backend
"""

import asyncio
import json
from datetime import datetime
from typing import Dict, Any, Optional, List
import httpx
from httpx import Timeout, Limits
import structlog

from ..config import settings

logger = structlog.get_logger(__name__)


class BitsaccoAPIClient:
    """Production-ready Bitsacco API client with retry logic and caching"""

    def __init__(self):
        self.base_url = settings.BITSACCO_API_URL.rstrip("/")
        self.api_key = settings.BITSACCO_API_KEY
        self.client: Optional[httpx.AsyncClient] = None

        # Request configuration
        self.timeout = Timeout(settings.BITSACCO_TIMEOUT)
        self.limits = Limits(max_keepalive_connections=5, max_connections=10)

        # Rate limiting and retry
        self.max_retries = 3
        self.retry_delay = 1.0

        # Health status
        self.is_healthy = False
        self.last_health_check = None

    async def initialize(self) -> None:
        """Initialize HTTP client with proper configuration"""
        try:
            self.client = httpx.AsyncClient(
                timeout=self.timeout,
                limits=self.limits,
                headers={
                    "Authorization": f"Bearer {self.api_key}",
                    "Content-Type": "application/json",
                    "User-Agent": f"Bitsacco-WhatsApp-Bot/{settings.VERSION}",
                    "Accept": "application/json",
                },
            )

            # Perform initial health check
            await self.health_check()

            logger.info("✅ Bitsacco API client initialized")

        except Exception as e:
            logger.error("❌ Failed to initialize Bitsacco API client", error=str(e))
            raise

    async def health_check(self) -> Dict[str, Any]:
        """Check API health and connectivity"""
        try:
            response = await self._make_request("GET", "/health")

            if response and response.get("status") == "ok":
                self.is_healthy = True
                self.last_health_check = datetime.utcnow()
                return {"status": "healthy", "api_status": "operational"}
            else:
                self.is_healthy = False
                return {"status": "unhealthy", "api_status": "degraded"}

        except Exception as e:
            self.is_healthy = False
            logger.error("Bitsacco API health check failed", error=str(e))
            return {"status": "error", "error": str(e)}

    async def get_user_by_phone(self, phone_number: str) -> Dict[str, Any]:
        """Get user information by phone number"""
        try:
            endpoint = f"/users/phone/{phone_number}"
            response = await self._make_request("GET", endpoint)

            if response:
                return {
                    "exists": True,
                    "user_id": response.get("id"),
                    "phone_number": response.get("phone_number"),
                    "name": response.get("name"),
                    "kyc_level": response.get("kyc_level", "basic"),
                    "account_status": response.get("status", "active"),
                }
            else:
                return {"exists": False, "message": "User not found"}

        except httpx.HTTPStatusError as e:
            if e.response.status_code == 404:
                return {"exists": False, "message": "User not found"}
            else:
                logger.error(
                    "Error fetching user by phone", error=str(e), phone=phone_number
                )
                raise

        except Exception as e:
            logger.error(
                "Unexpected error fetching user", error=str(e), phone=phone_number
            )
            raise

    async def send_otp(self, phone_number: str) -> Dict[str, Any]:
        """Send OTP to user's phone number"""
        try:
            payload = {
                "phone_number": phone_number,
                "action": "whatsapp_login",
                "channel": "sms",
            }

            response = await self._make_request("POST", "/auth/send-otp", json=payload)

            if response:
                return {
                    "success": True,
                    "message": "OTP sent successfully",
                    "expires_in": response.get("expires_in", 300),
                    "request_id": response.get("request_id"),
                }
            else:
                return {"success": False, "message": "Failed to send OTP"}

        except Exception as e:
            logger.error("Error sending OTP", error=str(e), phone=phone_number)
            return {"success": False, "message": f"Error: {str(e)}"}

    async def verify_otp(self, phone_number: str, otp_code: str) -> Dict[str, Any]:
        """Verify OTP code"""
        try:
            payload = {
                "phone_number": phone_number,
                "otp_code": otp_code,
                "action": "whatsapp_login",
            }

            response = await self._make_request(
                "POST", "/auth/verify-otp", json=payload
            )

            if response and response.get("valid"):
                return {
                    "success": True,
                    "user_id": response.get("user_id"),
                    "access_token": response.get("access_token"),
                    "expires_in": response.get("expires_in", 3600),
                }
            else:
                return {
                    "success": False,
                    "message": response.get("message", "Invalid OTP"),
                }

        except Exception as e:
            logger.error("Error verifying OTP", error=str(e), phone=phone_number)
            return {"success": False, "message": f"Error: {str(e)}"}

    async def get_balance(self, user_id: str) -> Dict[str, Any]:
        """Get user's wallet balance"""
        try:
            endpoint = f"/users/{user_id}/balance"
            response = await self._make_request("GET", endpoint)

            if response:
                return {
                    "success": True,
                    "btc_balance": response.get("btc_balance", 0.0),
                    "kes_balance": response.get("kes_balance", 0.0),
                    "usd_balance": response.get("usd_balance", 0.0),
                    "last_updated": response.get("last_updated"),
                    "wallet_address": response.get("wallet_address"),
                }
            else:
                return {"success": False, "message": "Failed to fetch balance"}

        except Exception as e:
            logger.error("Error fetching balance", error=str(e), user_id=user_id)
            return {"success": False, "message": f"Error: {str(e)}"}

    async def initiate_bitcoin_savings(
        self, phone_number: str, amount_kes: float
    ) -> Dict[str, Any]:
        """Initiate Bitcoin savings transaction via M-Pesa"""
        try:
            payload = {
                "phone_number": phone_number,
                "amount_kes": amount_kes,
                "payment_method": "mpesa",
                "transaction_type": "bitcoin_savings",
                "currency": "KES",
            }

            response = await self._make_request(
                "POST", "/transactions/savings/initiate", json=payload
            )

            if response:
                return {
                    "success": True,
                    "transaction_id": response.get("transaction_id"),
                    "amount_kes": amount_kes,
                    "estimated_btc": response.get("estimated_btc"),
                    "exchange_rate": response.get("exchange_rate"),
                    "mpesa_reference": response.get("mpesa_reference"),
                    "status": response.get("status", "pending"),
                    "expires_at": response.get("expires_at"),
                }
            else:
                return {"success": False, "message": "Failed to initiate savings"}

        except Exception as e:
            logger.error(
                "Error initiating Bitcoin savings",
                error=str(e),
                phone=phone_number,
                amount=amount_kes,
            )
            return {"success": False, "message": f"Error: {str(e)}"}

    async def get_transaction_history(
        self, user_id: str, limit: int = 10, offset: int = 0
    ) -> Dict[str, Any]:
        """Get user's transaction history"""
        try:
            params = {"limit": limit, "offset": offset, "order": "desc"}
            endpoint = f"/users/{user_id}/transactions"

            response = await self._make_request("GET", endpoint, params=params)

            if response:
                return {
                    "success": True,
                    "transactions": response.get("transactions", []),
                    "total": response.get("total", 0),
                    "has_more": response.get("has_more", False),
                }
            else:
                return {
                    "success": False,
                    "message": "Failed to fetch transaction history",
                }

        except Exception as e:
            logger.error(
                "Error fetching transaction history", error=str(e), user_id=user_id
            )
            return {"success": False, "message": f"Error: {str(e)}"}

    async def get_transaction_status(self, transaction_id: str) -> Dict[str, Any]:
        """Get status of specific transaction"""
        try:
            endpoint = f"/transactions/{transaction_id}/status"
            response = await self._make_request("GET", endpoint)

            if response:
                return {
                    "success": True,
                    "transaction_id": transaction_id,
                    "status": response.get("status"),
                    "amount_kes": response.get("amount_kes"),
                    "amount_btc": response.get("amount_btc"),
                    "created_at": response.get("created_at"),
                    "completed_at": response.get("completed_at"),
                    "blockchain_hash": response.get("blockchain_hash"),
                }
            else:
                return {"success": False, "message": "Transaction not found"}

        except Exception as e:
            logger.error(
                "Error fetching transaction status",
                error=str(e),
                transaction_id=transaction_id,
            )
            return {"success": False, "message": f"Error: {str(e)}"}

    async def update_user_preferences(
        self, user_id: str, preferences: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Update user preferences"""
        try:
            endpoint = f"/users/{user_id}/preferences"
            response = await self._make_request("PATCH", endpoint, json=preferences)

            if response:
                return {
                    "success": True,
                    "preferences": response.get("preferences"),
                    "updated_at": response.get("updated_at"),
                }
            else:
                return {"success": False, "message": "Failed to update preferences"}

        except Exception as e:
            logger.error("Error updating preferences", error=str(e), user_id=user_id)
            return {"success": False, "message": f"Error: {str(e)}"}

    async def _make_request(
        self,
        method: str,
        endpoint: str,
        json: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None,
    ) -> Optional[Dict[str, Any]]:
        """Make HTTP request with retry logic and error handling"""
        if not self.client:
            raise RuntimeError("API client not initialized")

        url = f"{self.base_url}{endpoint}"

        for attempt in range(self.max_retries + 1):
            try:
                logger.debug(
                    "Making API request",
                    method=method,
                    endpoint=endpoint,
                    attempt=attempt + 1,
                )

                response = await self.client.request(
                    method=method, url=url, json=json, params=params
                )

                # Log response details
                logger.debug(
                    "API response received",
                    status_code=response.status_code,
                    endpoint=endpoint,
                )

                # Handle different response codes
                if response.status_code == 200:
                    return response.json()
                elif response.status_code == 201:
                    return response.json()
                elif response.status_code == 204:
                    return {}
                elif response.status_code == 404:
                    return None
                elif response.status_code in [401, 403]:
                    logger.error(
                        "Authentication failed", status_code=response.status_code
                    )
                    raise httpx.HTTPStatusError(
                        message="Authentication failed",
                        request=response.request,
                        response=response,
                    )
                elif response.status_code >= 500:
                    # Server error - retry
                    if attempt < self.max_retries:
                        wait_time = self.retry_delay * (2**attempt)
                        logger.warning(
                            f"Server error, retrying in {wait_time}s",
                            status_code=response.status_code,
                            attempt=attempt + 1,
                        )
                        await asyncio.sleep(wait_time)
                        continue
                    else:
                        response.raise_for_status()
                else:
                    # Client error - don't retry
                    response.raise_for_status()

            except httpx.TimeoutException:
                if attempt < self.max_retries:
                    wait_time = self.retry_delay * (2**attempt)
                    logger.warning(
                        f"Request timeout, retrying in {wait_time}s",
                        endpoint=endpoint,
                        attempt=attempt + 1,
                    )
                    await asyncio.sleep(wait_time)
                    continue
                else:
                    logger.error("Request timeout after all retries", endpoint=endpoint)
                    raise

            except httpx.NetworkError as e:
                if attempt < self.max_retries:
                    wait_time = self.retry_delay * (2**attempt)
                    logger.warning(
                        f"Network error, retrying in {wait_time}s",
                        error=str(e),
                        attempt=attempt + 1,
                    )
                    await asyncio.sleep(wait_time)
                    continue
                else:
                    logger.error("Network error after all retries", error=str(e))
                    raise

            except Exception as e:
                logger.error(
                    "Unexpected error in API request", error=str(e), endpoint=endpoint
                )
                raise

        return None

    async def close(self) -> None:
        """Close HTTP client"""
        if self.client:
            await self.client.aclose()
            logger.info("Bitsacco API client closed")

    def __del__(self):
        """Cleanup on deletion"""
        if self.client:
            # Note: This won't work properly in async context
            # Proper cleanup should be done via close() method
            pass


# Alias for backward compatibility
BitsaccoAPIService = BitsaccoAPIClient
